core/src/test/java/org/elasticsearch/search/profile/ProfileResultTests.java

```diff
@@ -31,38 +34,86 @@
 import java.util.List;
 import java.util.Map;
 
+import static org.elasticsearch.common.xcontent.XContentHelper.toXContent;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertToXContentEquivalent;
+
 public class ProfileResultTests extends ESTestCase {
 
+    public static ProfileResult createTestItem(int depth) {
+        String type = randomAsciiOfLengthBetween(5, 10);
+        String description = randomAsciiOfLengthBetween(5, 10);
+        int timingsSize = randomIntBetween(0, 5);
+        Map<String, Long> timings = new HashMap<>(timingsSize);
+        for (int i = 0; i < timingsSize; i++) {
+            long time = randomNonNegativeLong() / timingsSize;
+            if (randomBoolean()) {
+                // also often use "small" values in tests
+                time = randomNonNegativeLong() % 10000;
+            }
+            timings.put(randomAsciiOfLengthBetween(5, 10), time); // don't overflow Long.MAX_VALUE;
+        }
+        int childrenSize = depth > 0 ? randomIntBetween(0, 1) : 0;
+        List<ProfileResult> children = new ArrayList<>(childrenSize);
+        for (int i = 0; i < childrenSize; i++) {
+            children.add(createTestItem(depth - 1));
+        }
+        return new ProfileResult(type, description, timings, children);
+    }
+
+    public void testFromXContent() throws IOException {
+        ProfileResult profileResult = createTestItem(2);
+        XContentType xContentType = randomFrom(XContentType.values());
+        boolean humanReadable = randomBoolean();
+        BytesReference originalBytes = toXContent(profileResult, xContentType, humanReadable);
+        try (XContentParser parser = createParser(xContentType.xContent(), originalBytes)) {
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+            ProfileResult parsed = ProfileResult.fromXContent(parser);
+            assertEquals(profileResult.getTime(), parsed.getTime());
```

Comment:
Can we check the parser in the try/catch block, and the parsed `ProfileResult` after that?

It's basically moving those two lines:
ProfileResult parsed = ProfileResult.fromXContent(parser);
assertEquals(profileResult.getTime(), parsed.getTime());

after the try block.